rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+# RFC 026: Work Enricher
+
+## Background
+
+Recently we have started ingesting archive data from Calm. An archive according to our model is a collection of works arranged hierarchically in a tree structure according to some path. For example, the works with paths `A`, `A/1`, `A/1/X`, `A/1/Y` and `A/2` would for the following tree:
+
+```
+A
+├── 1
+│   ├── X
+│   └── Y
+│
+└── 2
+```
+
+Our API allows us to display relations for some particular work using a number of fields: 
+
+* `parts` is an array containing the children of the work (i.e. `parts` for work `A` would contain `A/1` and `A/2`)
+
+* `partOf` is an array containing the ancestors of the work (i.e. `partOf` for work `A/1/X` would contain `A` and `A/1`)
+
+* `precededBy` is an array containing the siblings of the work earlier in sort order (i.e. `precededBy` for work `A/2` would contain `A/1`)
+
+* `succeededBy` is an array containing the siblings of the work later in sort order (i.e. `succeededBy` for work `A/1/X` would contain `A/1/Y`)
+
+Note that the works contained within any of these relation fields are partial (only a core number of fields).
+
+## Problem
+
+Currently we calculate work relations dynamically on the server during a request cycle in the API, by querying Elasticsearch given some work path: we first fetch the original work by ID, then sequentially send a group of follow up queries for the relations (using the Elasticsearch multi search functionality). This approach has a number of drawbacks:
+
+* The API performs unnecessary work which could be done up front to provide faster response times.
+
+* It is not simple to extend this functionality to display relation data on the multi-work API endpoint, with it currently limited to the single work endpoint (the naive method of one extra request per work would result in a great number of Elasticsearch queries within a single incoming HTTP request).
+
+* The fact that the relations are not indexed up front means we can not use related data within search queries (such as if we wanted to include the title of the root of the collection for any descendent works, for example).
+
+* The use of paths for encoding the structure of the tree couples relations tightly to Calm. There is upcoming use cases for relations beyond just archives, and on a conceptual level the goal of the pipeline should be to generate data which is abstracted away from any particular source system.
+
+We therefore need a method to generate relation data on works within the pipeline, in advance of indexing for the API.
+
+## Enricher
+
+In order to include relation data on the work model itself it is clear there needs to be an additional stage of the pipeline which is able to query relationships between works, and include data from related works onto the work itself. This proposed name for this stage is the enricher.
```

Comment:
I wonder if the name for this design should indicate that this is a kind of "relational enricher", as otherwise it's not clear that (whilst the task performed is still enrichment) it differs from eg an image enricher that doesn't analyse relationships but instead performs ML inference.

Reply:
fwiw this distinction gets a bit trickier with the concepts enricher where we're planning to do both

Reply:
Perhaps we don't have to include the name in the services then - it would just be good to have one term for this design which uniquely identifies it eg "the works enricher is a relational enricher", "the concepts enricher is a relational and inference enricher"

Reply:
Not really important but I think just enricher is a bit vague in that almost every step of the pipeline is enriching. In fact, the whole pipeline could be called enricher. I would prefer a name that hints in what way it's enriching

Reply:
We're deep in the semantics here and I don't want to distract from the actual work at hand, but I agree with @jamesgorrie.

I see enrichers as distinct from other pipeline services in that works/images/concepts/whatever should come out of an enricher with _new_ data which doesn't appear anywhere in the original sources.

That might be because we're fetching information from third parties (eg the concepts enricher), or because we're doing some internal analysis (eg extracting image features and colour palettes), but generally, an enrichment step should be something that adds new information.

Reply:
The description of what enriching from @harrisonpim makes complete sense to me. But that makes me question whether this is really doing any enriching :thinking:, its more copying existing data from elsewhere. "Denormaliser" doesn't really have the same ring to it though.

Reply:
☝️ fits with @jamesgorrie's comment below about the risk of over-generalising what an enricher is. I don't see anything wrong with calling this a 'denormaliser' tbh, especially if its functionality is going to appear in a few places across the pipeline.